core/commands/dag/dag.go

```diff
@@ -241,3 +279,435 @@ var DagResolveCmd = &cmds.Command{
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		// !!!NOTE!!! - We take .GCLock, *not* .PinLock. The way the current stack is
+		// built makes it impossible to take both a GC and a Pinning lock, regardless
+		// of order. Doing so can potentially deadlock if another thread starts GC
+		// ( this is not theoretical - it's been reliably simulated with the insertion
+		// of strategic sleeps()s and hand-triggering GC )
+		//
+		// See api.Pin().Add() vs node.Pinning.Pin() comment further down
+		unlocker := node.Blockstore.GCLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+		showProgress, _ := req.Options[progressOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		var counts ObjectCounts
+		go importWorker(req, &res, &api, &counts, retCh)
+
+		var roots *expectedRootsSeen
+
+		if !showProgress {
+			done := <-retCh
+			if done.err != nil {
+				return done.err
+			}
+			roots = done.roots
+		} else {
+
+			progressTicker := time.NewTicker(500 * time.Millisecond)
+			defer progressTicker.Stop()
+
+		ImportLoop:
+			for {
+				select {
+				case ret := <-retCh:
+					if ret.err != nil {
+						return ret.err
+					}
+					roots = ret.roots
+
+					progressTicker.Stop()
+					if err := res.Emit(&CarImportOutput{ObjectCounts: counts}); err != nil {
+						return err
+					}
+					os.Stderr.WriteString("\n")
+
+					break ImportLoop
+				case <-progressTicker.C:
+					if err := res.Emit(&CarImportOutput{ObjectCounts: counts}); err != nil {
+						return err
+					}
+				case <-req.Context.Done():
+					return req.Context.Err()
+				}
+			}
+		}
+
+		// opportunistic pinning: try whatever sticks
+		if doPinRoots {
+
+			for c, seen := range *roots {
+
+				// We need to re-retrieve a block, convert it to ipld, and feed it
+				// to the Pinning interface, sigh...
+				//
+				// If we didn't have the deadlocking-gclock/pinlock problem we could
+				// use the Api directly like so (though internally it does the same):
+				//
+				// // not ideal, but the pinning api takes only paths :(
+				// rp := path.NewResolvedPath(
+				// 	ipfspath.FromCid(c),
+				// 	c,
+				// 	c,
+				// 	"",
+				// )
+				//
+				// if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
+
+				ret := RootMeta{Cid: &c, PresentInCar: seen}
```

Comment:
What if `seen` is `false` but the CID exists in the blockstore already. I _think_ this would mean that it gets pinned regardless of the "malformed" CAR. Should the next block of if/else's be contained in an `if seen`?

Reply:
This is correct/by design: https://github.com/ipfs/go-ipfs/blob/03fe9bde87462aa11775bf6e8361a56fd967f89d/test/sharness/t0054-dag-car-import-export.sh#L83-L86

Reply:
this goes back to the question of "malformed" then, are you sure you want to have a lose edge like this? might be better to be a little more strict about the contents of a car - "you don't have the root you say you do, bad luck", instead you are opening up edges that could become important features for some in the future when we decide to be more strict with a future CAR format, or could lead to unexpected behaviour for others.